src/game/entities/Control.ts

```diff
@@ -37,6 +53,25 @@ export default class Control {
   init(): void {
     document.addEventListener('keydown', this.handleKeyPressed);
     document.addEventListener('keyup', this.handleKeyReleased);
+    window.addEventListener('gamepadconnected', this.handleGamepadConnected);
+    window.addEventListener('gamepaddisconnected', this.handleGamepadDisconnected);
+  }
+
+  update(): void {
+    if (!this.gamepad) {
+      return;
+    }
+
+    Object.keys(this.registeredGamepadButtons).forEach((button) => {
+      const key = this.registeredGamepadButtons[+button];
+
+      if (navigator.getGamepads()[this.gamepad?.index || 0]?.buttons[+button].pressed) {
```

Comment:
Может лучше сразу обращаться к `this.gamepad`, и не запрашивать `navigator.getGamepads()` на каждый `update` ?

Также у меня тут ошибка после нажатия кнопки джойстика: https://i.imgur.com/Jc0aX8c.png (у меня для Defender Game Master G2 в `gamepad.buttons` 10 кнопок, т.е. кнопки с индексом 14 нет). Не знаю как делается адаптация для всех джойстиков, может при подключении считывать количество кнопок и от этого плясать, но ещё нужно индексы как-то сопоставить с самими кнопками, чтоб они не были рандомными (ну это наверно невозможно если у разных джойстиков кнопки рандомно нумеруются)

Reply:
Точно не хочется видеть всё это в условии
```suggestion
      const currentGamepad = navigator.getGamepads()[this.gamepad?.index || 0]

      if (currentGamepad?.buttons[+button].pressed) {
```

Reply:
@andreyvolokitin единственный способ поддержать таким образом все вариации геймпадов — заставить игрока биндить всё по факту подключения. То есть подключаешь геймпад и тебя заставляют забиндить все кнопки

Reply:
@andreyvolokitin это первое что пришло в голову. Но объект `this.gamepad` не обновляется.
А по кнопкам да, единственный способ вижу как @kotosha-real предложил.

@kotosha-real вынес из условия